controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go

```diff
@@ -0,0 +1,244 @@
+	if _, _, err := vp.decoder.Decode(cp.Spec.ProviderConfig.Raw, nil, cpConfig); err != nil {
+		return nil, errors.Wrapf(err, "could not decode providerConfig of controlplane '%s'", util.ObjectName(cp))
+	}
+
+	// Decode infrastructureProviderStatus
+	infraStatus := &openstackv1alpha1.InfrastructureStatus{}
+	if _, _, err := vp.decoder.Decode(cp.Spec.InfrastructureProviderStatus.Raw, nil, infraStatus); err != nil {
+		return nil, errors.Wrapf(err, "could not decode infrastructureProviderStatus of controlplane '%s'", util.ObjectName(cp))
+	}
+
+	// Get credentials
+	credentials, err := internal.GetCredentialsForNamespaceAndName(ctx, vp.client, cp.Spec.SecretRef.Namespace, cp.Spec.SecretRef.Name)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get service account from secret '%s/%s'", cp.Spec.SecretRef.Namespace, cp.Spec.SecretRef.Name)
+	}
+
+	// Get config chart values
+	return getConfigChartValues(cpConfig, infraStatus, cp, credentials, cluster)
+}
+
+// GetControlPlaneChartValues returns the values for the control plane chart applied by the generic actuator.
+func (vp *valuesProvider) GetControlPlaneChartValues(
+	ctx context.Context,
+	cp *extensionsv1alpha1.ControlPlane,
+	cluster *extensionscontroller.Cluster,
+	checksums map[string]string,
+) (map[string]interface{}, error) {
+	// Decode providerConfig
+	cpConfig := &openstackv1alpha1.ControlPlaneConfig{}
+	if _, _, err := vp.decoder.Decode(cp.Spec.ProviderConfig.Raw, nil, cpConfig); err != nil {
+		return nil, errors.Wrapf(err, "could not decode providerConfig of controlplane '%s'", util.ObjectName(cp))
+	}
+
+	// Get CCM chart values
+	return getCCMChartValues(cpConfig, cp, cluster, checksums)
+}
+
+// GetControlPlaneShootChartValues returns the values for the control plane shoot chart applied by this actuator.
+func (vp *valuesProvider) GetControlPlaneShootChartValues(
+	context.Context,
+	*extensionsv1alpha1.ControlPlane,
+	*extensionscontroller.Cluster,
+) (map[string]interface{}, error) {
+	return nil, nil
+}
+
+// getConfigChartValues collects and returns the configuration chart values.
+func getConfigChartValues(
+	cpConfig *openstackv1alpha1.ControlPlaneConfig,
+	infraStatus *openstackv1alpha1.InfrastructureStatus,
+	cp *extensionsv1alpha1.ControlPlane,
+	c *internal.Credentials,
+	cluster *extensionscontroller.Cluster,
+) (map[string]interface{}, error) {
+
+	// Collect config chart values
+	return map[string]interface{}{
+		"kubernetesVersion": cluster.Shoot.Spec.Kubernetes.Version,
+		"domainName":        c.DomainName,
+		"tenantName":        c.TenantName,
+		"username":          c.Username,
+		"password":          c.Password,
+		"lbProvider":        cpConfig.LoadBalancerProvider,
+		"floatingNetworkID": infraStatus.Networks.FloatingPool.ID,
+		"subnetID":          getSubnetID(infraStatus),
+		"dhcpDomain":        cpConfig.DHCPDomain,
+		"requestTimeout":    cpConfig.RequestTimeout,
+	}, nil
+}
+
+// TODO should we return error in case the subnetID is nil?
```

Comment:
Yes, we should. The logic for AWS and Alicloud is similar, see how it's implemented there.